### PR TITLE
[BugFix] Only support olap tables/materaizlied views without related mvs for non lock optimization

### LINKED ISSUE #4777: Refactor return type in Expression Framework
## Refactoring request

now it's hard to handle a wrong result in our Expression Framework. 

if `ExprContext` could return a Column with a Status code. we could handle execution errors in functions more convenient. such as return a runtime error (BinaryColumn greater than 4G)

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #4777

Why I'm doing:

After PR https://github.com/StarRocks/starrocks/pull/34569/files, olap/native olaptable/materialized view can use non lock optimization, but query tables' related mvs have not 
copy its metadata, so optimization may be corrupted in the concurrent scenes.

What I'm doing:

- Support copy-only for queries' related materialized views.
- Add `skip_whole_phase_lock_mv_limit` to control strategies:
```
    /**
     * To avoid too many related materialized view causing too much fe memory and decreasing performance, set N
     * to determine which strategy you choose:
     *  N <0      : always use non lock optimization and no copy related materialized views which
     *      may cause metadata concurrency problem but can reduce many lock conflict time and metadata memory-copy consume.
     *  N = 0    : always not use non lock optimization
     *  N > 0    : use non lock optimization when related mvs's num <= N, otherwise don't use non lock optimization
     */
    @ConfField(mutable = true)
    public static int skip_whole_phase_lock_mv_limit = 5;
```

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 1]
```diff
@@ -1115,6 +1115,17 @@ public class Config extends ConfigBase {
...
+    /**
+     * To avoid too many related materialized view causing too much fe memory and decreasing performance, set N
+     * to determine which strategy you choose:
+     *  N <0      : always use non lock optimization and no copy related materialized views which
+     *      may cause metadata concurrency problem but can reduce many lock conflict time and metadata memory-copy consume.
+     *  N = 0    : always not use non lock optimization
+     *  N > 0    : use non lock optimization when related mvs's num <= N, otherwise don't use non lock optimization
+     */
+    @ConfField(mutable = true)
+    public static int non_lock_optimization_max_related_materialized_view_num = 5;
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:1127]: Put this config item into GlobalVariables?    Config command is hard to remember and write.

use a clear name? for an example
```
skip_whole_phase_lock_mv_limit
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java [hunk 2]
```diff
@@ -310,8 +381,15 @@ private void preprocessMv(MaterializedView mv, Set<Table> queryTables,
             refTableUpdatedPartitionNames = mv.getUpdatedPartitionNamesOfTable(refBaseTable, true);
         }
 
+        MaterializedView copiedMV = mv;
+        if (context.getQueryTables() != null) {
+            // If query tables are set which means use related mv for non lock optimization,
+            // copy mv's metadata into a ready-only object.
+            copiedMV = copyOnlyMaterializedView(mv);
+        }
+
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java:390]:
```java
        MaterializedView copiedMV = mv;
        if (context.getQueryTables() != null) {
            // If query tables are set which means use related mv for non lock optimization,
            // copy mv's metadata into a ready-only object.
            copiedMV = copyOnlyMaterializedView(mv);
        }
```

suggest to write like this

```java
MaterializedView copiedMV=Optional.ofNullable((context.getQueryTables()).map{a->copyOnlyMaterializedView(mv)}.elseGet(mv);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java [hunk 3]
```diff
@@ -71,40 +74,143 @@ public class MvRewritePreprocessor {
...
+            try {
+                Set<MaterializedView> relatedMVs =
+                        getRelatedMVs(connectContext, queryTables, context.getOptimizerConfig().isRuleBased());
+                Set<Pair<MaterializedView, MvPlanContext>> validMVs = filterValidMVs(connectContext, relatedMVs);
+                prepareRelatedMVs(queryTables, validMVs);
+            } catch (Exception e) {
+                // MV's prepare should not affect query's process, this may be caused by MV's concurrent process.
+                List<String> tableNames = queryTables.stream().map(Table::getName).collect(Collectors.toList());
+                LOG.warn("Prepare query tables {} for mv failed: {}", tableNames, e);
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java:99]: Why not throw exception, what if exception happens

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java [hunk 4]
```diff
@@ -71,40 +74,143 @@ public class MvRewritePreprocessor {
...
+            }
+        }
+    }
+
+    private MaterializedView copyOnlyMaterializedView(MaterializedView mv) {
+        // TODO: add read lock?
+        // Query will not lock dbs in the optimizer stage, so use a shadow copy of mv to avoid
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java:106]: shallow copy

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java (hunk 3 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java:104]: throw exception just like before.

**Review by satanson — DISMISSED** [review]

**Review by satanson — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 5]
```diff
@@ -131,14 +130,18 @@ public OptExpression optimize(ConnectContext connectContext,
                                   ColumnRefSet requiredColumns,
                                   ColumnRefFactory columnRefFactory) {
         prepare(connectContext, logicOperatorTree, columnRefFactory);
-        context.setUpdateTableId(updateTableId);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:134]: Why remove this?

**Review by packy92 — APPROVED** [review]

**Pull request merged**